docs/architecture-decision-records/2023-01-initial-architecture.md

```diff
@@ -20,19 +20,127 @@ The modules should have access to computer resources - CPU, memory, storage, and
 we want to enforce limits on resource usage, to ensure the Station is unobtrusive and does not
 interfere with the actual work performed on the computer.
 
-Last but not least, we want our runtime to provide high-level APIs for module authors, exposing
-primitives for peer-to-peer networking via libp2p, content fetching and publishing via
-IPFS & Filecoin, and so on.
+We want our runtime to provide high-level APIs for module authors, exposing primitives for
+peer-to-peer networking via libp2p, content fetching and publishing via IPFS & Filecoin, and so on.
 
 We want the ability to share the same underlying libp2p and IPFS node by multiple Station modules,
 to avoid the overhead of duplicating connections to the same network peers.
 
+Modules should be capable of handling high concurrency, leveraging modern design patterns like
+asynchronous execution and non-blocking I/O.
+
 ## Options Considered
 
+See also [Station Runtime
+Research](https://www.notion.so/pl-strflt/2023-01-Station-Runtime-Research-c45c61a9397241bba98f0d67bafe4e5d)
+on Notion for additional information.
+
 <!--
 What are the different options we considered? What are their pros & cons?
 -->
 
+### Virtualization (Docker)
+
+Virtualization offers several great benefits:
+
+1. Sandboxing
+2. Resource usage limit
+3. Modules can be written in any language, composed from multiple heterogenous components
+   (e.g. nginx reverse proxy + Node.js app)
+
+However, we ruled out virtualization-based solutions for the following reasons:
+
+  1. Entry-level computers may not support CPU instructions required for virtualization.
+  2. Often, virtualization must be enabled in BIOS settings.
+  3. Windows Home edition does not support Microsoft's virtualization framework Hyper-V.
+  4. Sharing the same IPFS & libp2p node between multiple container instances is difficult.
+
+### Deno
+
+_https://deno.land_
+
+Deno offers JS & WASM runtime with filesystem and networking APIs, including capabilities-based
+sanboxing.
+
+Benefits:
+
+- Full support for both JavaScript/TypeScript and WebAssembly
+- Filesystem and network access is sandboxed out of the box
+- The engine and platform APIs are designed for asynchrony and non-blocking I/O
+- WebAPI-compliant Fetch API implementation
+- WebAPIs can be easily called from Rust via [wasm-bindgen](https://crates.io/crates/wasm-bindgen) and [web-sys](https://crates.io/crates/web-sys) crates.
+
+Issues:
+
+- Missing documentation for embedders
+
+- Static JS imports are not sandboxed. Add this line to your TS script `import * as google from 'https://google.com/'` - it will execute even if network access is not allowed.
```

Comment:
ouch! Can you run `deno bundle` to create a script that will have this `import` already inlined? I'm quite sure afterwards there will be no more network calls.

Reply:
Yeah, if we wanted to use Deno, then we should require modules to provide us a single file with the bundle of all source code.

Then we would have to implement parse that code to ensure it does not contain any static imports.

I think that should be easily doable 👍🏻

Reply:
If we use deno, we could also automate the processes of using `deno build` as part of importing/updating the module